crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+    util::BufferInitDescriptor, BufferDescriptor, BufferUsages, CommandEncoderDescriptor,
+    DownlevelFlags, COPY_BUFFER_ALIGNMENT,
+};
+
+use crate::{
+    mesh::{GpuMesh, Indices, Mesh, MeshVertexBufferLayouts},
+    render_asset::{prepare_assets, ExtractedAssets},
+    render_resource::Buffer,
+    renderer::{RenderAdapter, RenderDevice, RenderQueue},
+    Render, RenderApp, RenderSet,
+};
+
+/// A plugin that provides memory management services for mesh data.
+pub struct MeshAllocatorPlugin;
+
+/// Manages the assignment of mesh data to hardware buffers.
+///
+/// The Bevy renderer tries to pack vertex and index data for multiple meshes
+/// together so that multiple meshes can be drawn back-to-back without any
+/// rebinding. This resource manages these buffers.
+///
+/// Within each slab, or hardware buffer, the underlying allocation algorithm is
+/// [`offset-allocator`], a Rust port of Sebastian Aaltonen's hard-real-time C++
+/// [`OffsetAllocator`]. Slabs start small and then grow as their contents fill
+/// up, up to a maximum size limit. To reduce fragmentation, vertex and index
+/// buffers that are too large bypass this system and receive their own buffers.
+///
+/// The [`MeshAllocatorSettings`] allows you to tune the behavior of the
+/// allocator for better performance with your application. Most applications
+/// won't need to change the settings from their default values.
+#[derive(Resource)]
+pub struct MeshAllocator {
+    /// Holds all buffers and allocators.
+    slabs: HashMap<SlabId, Slab>,
+
+    /// Maps a layout to the slabs that hold elements of that layout.
+    ///
+    /// This is used when allocating, so that we can find the appropriate slab
+    /// to place an object in.
+    slab_layouts: HashMap<ElementLayout, Vec<SlabId>>,
+
+    /// Maps mesh asset IDs to the ID of the slabs that hold their vertex data.
+    mesh_id_to_vertex_slab: HashMap<AssetId<Mesh>, SlabId>,
+
+    /// Maps mesh asset IDs to the ID of the slabs that hold their index data.
+    mesh_id_to_index_slab: HashMap<AssetId<Mesh>, SlabId>,
+
+    /// The next slab ID to assign.
+    next_slab_id: SlabId,
+
+    /// Whether we can pack multiple vertex arrays into a single slab on this
+    /// platform.
+    ///
+    /// This corresponds to [`DownlevelFlags::BASE_VERTEX`], which is unset on
+    /// WebGL 2. On this platform, we must give each vertex array its own
+    /// buffer, because we can't adjust the first vertex when we perform a draw.
+    general_vertex_slabs_supported: bool,
+}
+
+/// Tunable parameters that customize the behavior of the allocator.
+///
+/// Generally, these parameters adjust the tradeoff between memory fragmentation
+/// and performance. You can adjust them as desired for your application. Most
+/// applications can stick with the default values.
+#[derive(Resource)]
+pub struct MeshAllocatorSettings {
+    /// The minimum size of a slab (hardware buffer), in bytes.
+    ///
+    /// The default value is 1 MB.
```

Comment:
MB or MiB?

Reply:
I'm not inclined to indulge SI's attempt to redefine terms that everyone knows :)

Reply:
MiB is base 2 though, instead of base 10, which seems like a better fit for memory sizes to me :)

Reply:
MB is base 2 for everyone except SI. I don't want to indulge SI's attempt to redefine MB as base 10 for silly reasons.

Reply:
I don't like the use of MiB but I changed it anyway.